src/core/net/srp_replication.cpp

```diff
@@ -0,0 +1,2591 @@
+{
+    Settings::SrpReplicationInfo settingsInfo;
+
+    switch (mState)
+    {
+    case kStateDiscovery:
+        VerifyOrExit(mDiscoveryEndTime <= TimerMilli::GetNow());
+        break;
+
+    case kStateDisabled:
+    case kStateRunning:
+        ExitNow();
+    }
+
+    // Check if any of the marked sessions is still not finished
+    // its sync and if so we wait for it to finish. If all are
+    // finished (or errored) we can start SRP server.
+
+    for (const Session &session : mSessions)
+    {
+        if (session.IsMarkedForSyncBeforeServerStart() && (session.GetPhase() != Session::kRoutineOperation) &&
+            !session.IsMarkedAsErrored())
+        {
+            ExitNow();
+        }
+    }
+
+    if (mDomainName.IsNull())
+    {
+        SuccessOrAssert(mDomainName.Set(mDefaultDomainName));
+        otLogInfoSrp("[srpl] no domain discovered, using default domain: %s", mDomainName.AsCString());
+    }
+
+    if (!mHasDatasetId)
+    {
+        // Select dataset ID if not set (e.g., in case of first/only
+        // SRPL peer). The highest byte (MSB) of the dataset ID is used
+        // as seq number in Thread Network Data SRP service entry. The
+        // last used seq number is persisted in non-volatile settings.
+        // If we are selecting dateset ID we ensure to set its seq
+        // number one after the last used value.
+
+        mDatasetId.GenerateRandom();
+        mHasDatasetId = true;
+
+        if (Get<Settings>().Read(settingsInfo) == kErrorNone)
+        {
+            mDatasetId.SetSeqNumber(settingsInfo.GetSeqNumber() + 1);
+        }
+
+        otLogInfoSrp("[srpl] Selected dataset id %s", mDatasetId.ToString().AsCString());
+    }
+
+    StartDnssdAdvertisement();
+    SetState(kStateRunning);
+
+    SuccessOrAssert(Get<Server>().SetAnycastModeSequenceNumber(mDatasetId.GetSeqNumber()));
+    Get<Server>().Enable();
+
+    settingsInfo.SetSeqNumber(mDatasetId.GetSeqNumber());
+    IgnoreError(Get<Settings>().Save(settingsInfo));
+
+exit:
+    return;
+}
+
+void Srpl::StartDnssdAdvertisement(void)
+{
+    static constexpr uint16_t kTxtDataSize   = 200 + Dns::Name::kMaxNameSize;
+    static constexpr uint8_t  kNumTxtEntries = 4;
+
+    uint8_t                        txtDataBuffer[kTxtDataSize];
+    MutableData<kWithUint16Length> txtData;
+    Dns::TxtEntry                  entries[kNumTxtEntries];
+    PeerId::HexString              idString;
+    DatasetId::HexString           datasetString;
+
+    // Prepare the TXT record data to register SRPL DNS-SD service.
+
+    txtData.Init(txtDataBuffer, sizeof(txtDataBuffer));
+
+    entries[0].mKey         = kTxtDataKeyDomain;
+    entries[0].mValue       = reinterpret_cast<const uint8_t *>(mDomainName.AsCString());
+    entries[0].mValueLength = StringLength(mDomainName.AsCString(), Dns::Name::kMaxNameSize);
+
+    entries[1].mKey         = kTxtDataKeyAllowsJoin;
+    entries[1].mValue       = kYesString;
+    entries[1].mValueLength = sizeof(kYesString);
```

Comment:
Just curious, is there a use case of advertising `join=no`?

Reply:
I don't think there is any use case for it when SRPL is used in Thread and part of a Thread BR, but perhaps there are use-cases if we have stand-alone SRPL providers.